algorithms/linfa-trees/src/extra_trees/hyperparameters.rs

```diff
@@ -0,0 +1,97 @@
+use crate::decision_trees::DecisionTreeParams;
+use linfa::{
+    error::{Error, Result},
+    Float, Label,
+};
+
+#[cfg(feature = "serde")]
+use serde_crate::{Deserialize, Serialize};
+
+/// The set of hyperparameters that can be specified for fitting a
+/// [extra trees](struct.ExtraTrees.html).
+///
+/// ### Example
+///
+/// ```rust
+/// use linfa_trees::{ExtraTrees, DecisionTree, SplitQuality};
+/// use linfa_datasets::iris;
+/// use linfa::prelude::*;
+///
+/// // Initialize the default set of parameters
+/// let decision_tree_params = DecisionTree::params();
+/// let params = ExtraTrees::params();
+/// // Set the parameters to the desired values
+/// let decision_tree_params = decision_tree_params.split_quality(SplitQuality::Entropy).max_depth(Some(5)).min_weight_leaf(2.).random_split(true);
+/// let params = params.decision_tree_params(decision_tree_params);
+///
+///
+/// // Load the data
+/// let (train, val) = linfa_datasets::iris().split_with_ratio(0.9);
+/// // Fit the decision tree on the training data
+/// let trees = params.fit(&train).unwrap();
+/// // Predict on validation and check accuracy
+/// let val_accuracy = trees.predict(&val).confusion_matrix(&val).unwrap().accuracy();
+/// assert!(val_accuracy > 0.99);
+/// ```
+///
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+#[derive(Clone, Copy, Debug)]
+pub struct ExtraTreesParams<F, L> {
+    pub decision_tree_params: DecisionTreeParams<F, L>,
+    pub num_estimators: usize,
+    pub max_features: Option<usize>,
```

Comment:
since `max_features` is used explicitly during the fitting of the constituent trees I think it makes more sense to move max_features to the `DecisionTreeParams`, so there's less magic in this implementation.

Reply:
Ooh yes, this was actually something I wanted to discuss! I'm not sure if there's a way to circumvent this problem, but in `fit` I don't have write access for any of the provided parameters (which makes sense) because I'm given a shared reference `&self`. But, the default for `max_features` is dependent on the number of features in the `dataset`. So, I ended up passing it in as an argument but stylistically I'm not a fan of that, especially since it never changes. 

If this move to `DecisionTreeParams` is to avoid the extra argument, then I think it won't actually serve that purpose. I can technically move max_features into `DecisionTreeParams` with little code change but I was wondering if conceptually it should belong there? scikit-learn seems to use it as a parameter for Extra Trees and not Decision Trees.

Reply:
My reasoning for moving it to `DecisionTreeParams` is that it's really explicitly a modification to how individual decision trees perform and by keeping it on in `ExtraTreeParams` we prevent an end user from using a single `DecisitionTree` with `max_features` in the same stream-lined way as the other hyperparameters, even though this functionality is supported.

My thinking is:
1. Add `max_features` to `DecisionTreeParams`
2. Keep it in `ExtraTreeParams` as well, and overwrite user-provided `max_features` in the internal `DecisionTreeParams` similar to how random_split is overwritten

We can discuss more today though